libcudacxx/include/cuda/std/detail/libcxx/include/__algorithm/comp.h

```diff
@@ -0,0 +1,57 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCUDACXX___ALGORITHM_COMP_H
+#define _LIBCUDACXX___ALGORITHM_COMP_H
+
+#ifndef __cuda_std__
+#  include <__config>
+#endif // __cuda_std__
+
+#include "../__type_traits/integral_constant.h"
+#if defined(_LIBCUDACXX_HAS_STRING)
+#  include "../__type_traits/predicate_traits.h"
+#endif // _LIBCUDACXX_HAS_STRING
+
+#if defined(_CCCL_COMPILER_NVHPC) && defined(_CCCL_USE_IMPLICIT_SYSTEM_DEADER)
+#  pragma GCC system_header
+#else // ^^^ _CCCL_COMPILER_NVHPC ^^^ / vvv !_CCCL_COMPILER_NVHPC vvv
+_CCCL_IMPLICIT_SYSTEM_HEADER
+#endif // !_CCCL_COMPILER_NVHPC
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+struct __equal_to
+{
+  template <class _T1, class _T2>
+  _LIBCUDACXX_NODISCARD_EXT _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 bool
+  operator()(const _T1& __lhs, const _T2& __rhs) const noexcept(noexcept(__lhs == __rhs))
+  {
+    return __lhs == __rhs;
+  }
+};
+
+#if defined(_LIBCUDACXX_HAS_STRING)
+template <class _Lhs, class _Rhs>
+struct __is_trivial_equality_predicate<__equal_to, _Lhs, _Rhs> : true_type
+{};
+#endif // _LIBCUDACXX_HAS_STRING
+
+struct __less
+{
+  template <class _Tp, class _Up>
+  _LIBCUDACXX_NODISCARD_EXT _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 bool
+  operator()(const _Tp& __lhs, const _Up& __rhs) const noexcept(noexcept(__lhs < __rhs))
+  {
+    return __lhs < __rhs;
```

Comment:
suggestion: this and other functions are marked as `__host__ __device__`. I think we'll need to guard these functions with `nv_exec_check_disable`.

Reply:
So in principle I agree, but we would have to add them in a *lot* of places, so I would rather do that as a followup

Reply:
I'm a bit confused why we [don't care](https://github.com/NVIDIA/cccl/pull/627#discussion_r1375178652) about exceptions but add noexcept while we do care about exec check disable but don't add it. If you decide to add it as a follow up, please file an issue and link it here.

Reply:
From my point of view, noexcept specifiers are QoI and do not affect correctness in any way. The program will not become incorrect if I add a `noexcept` specifier nor will a incorrect program start to compile. The only potential change is that a program that should compile now does, because we properly propagate that noexcept specifier and a user might select on that.

The `nv_exec_check_disable` however potentially changes correctness of the program. A previously not-compiling incorrect program will now become a "fails at runtime" program. So this is something I would rather pull out into a different PR   because it does fundamentally change functionality

Reply:
Thank you for elaborating! 

> The program will not become incorrect if I add a noexcept specifier

Same is true with `_LIBCUDACXX_DISABLE_EXEC_CHECK`. A program will not become incorrect if you add it. 

> A previously not-compiling incorrect program will now become a "fails at runtime" program. 

The issue is that correct program also doesn't compile:
```cpp
struct D {};
__device__ bool operator==(D, D) { return true; }
__global__ void kernel() {
  cuda::std::__equal_to{}(D{}, D{});
}
//  error: calling a __device__ function("operator ==(    ::D,     ::D)") from a __host__ __device__ function("operator()") is not allowed
```
Please, file an issue to add `_LIBCUDACXX_DISABLE_EXEC_CHECK` and link it here.

Reply:
Filed #648